engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt

```diff
@@ -67,21 +67,43 @@ abstract class FhirSyncWorker(appContext: Context, workerParams: WorkerParameter
 
     val synchronizer =
       FhirSynchronizer(
-        applicationContext,
         getFhirEngine(),
         Uploader(dataSource),
         DownloaderImpl(dataSource, getDownloadWorkManager()),
         getConflictResolver(),
+        FhirEngineProvider.getFhirDataStore(applicationContext),
       )
 
     val job =
       CoroutineScope(Dispatchers.IO).launch {
-        synchronizer.syncState.collect {
-          // now send Progress to work manager so caller app can listen
-          setProgress(buildWorkData(it))
-
-          if (it is SyncJobStatus.Finished || it is SyncJobStatus.Failed) {
-            this@launch.cancel()
+        val fhirDataStore = FhirEngineProvider.getFhirDataStore(applicationContext)
+        synchronizer.syncState.collect { syncJobStatus ->
+          val uniqueWorkerName = inputData.getString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY)
+          when (syncJobStatus) {
+            is SyncJobStatus.Started,
+            is SyncJobStatus.InProgress, -> {
+              setProgress(buildWorkData(syncJobStatus))
+              // As the currentJobStatus is in progress, update the terminal state to unknown.
+              uniqueWorkerName?.let {
+                fhirDataStore.updateSyncJobTerminalState(it, SyncJobStatus.Unknown)
+              }
+            }
+            is SyncJobStatus.Finished,
+            is SyncJobStatus.Failed, -> {
+              // While creating periodicSync request if
+              // putString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY, uniqueWorkName) is not present,
+              // then inputData.getString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY) can be null.
+              if (uniqueWorkerName == null) {
+                setProgress(buildWorkData(syncJobStatus))
+              } else {
+                fhirDataStore.updateSyncJobTerminalState(uniqueWorkerName, syncJobStatus)
+                fhirDataStore.updateLastSyncJobStatus(uniqueWorkerName, syncJobStatus)
+              }
+              cancel()
+            }
+            is SyncJobStatus.Unknown -> {
+              Timber.e("$syncJobStatus is received.")
+            }
           }
```

Comment:
we only write to the fhirstore when in a terminal state, which simplifies the logic a lot
```suggestion
          when (syncJobStatus) {
            is SyncJobStatus.Finished,
            is SyncJobStatus.Failed, -> {
              // While creating periodicSync request if
              // putString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY, uniqueWorkName) is not present,
              // then inputData.getString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY) can be null.
              if (uniqueWorkerName != null) {
                fhirDataStore.writeTerminalSyncJobStatus(uniqueWorkerName, syncJobStatus)
              }
              cancel()
            }
            else -> setProgress(buildWorkData(syncJobStatus))
          }
```